### PR TITLE
User-set proposal titles, part 4: Author-setting

### LINKED ISSUE #840: User-set title on chain entities
### Summary

It should become possible to set a title on Chain Entities, in which case the current title (e.g. for a treasury proposal “Proposed spend: 50,000 EDG to kour8…”) would change to a manually set title. If no title has been set, on the on-chain governance pages, “No title set” should be displayed in a lighter color instead.

The title column would be used for:
* Referenda
* Democracy Proposals
* Treasury Proposals

The title column would not be used for:
* Signaling Proposals
* Council Motions
* Non-Substrate proposals.

### Expected Behavior

- [ ] Add a carat menu (similar to discussion rows) to referenda, democracy proposals, and treasury proposals. If it is straightforward, this should be placed both on each row in the proposal listing page, and on the view_proposal page. It should be hidden for other unsupported proposal types.
- [ ] The carat menu should offer one option, Set title, which is only visible to Admins, Moderators, and the `proposer` of the proposal. That menu option should pop open a modal, with the title “Set title” and an input box for the title. If a title is submitted by a valid user, it should be saved to the database. The modal should handle errors correctly (e.g. by showing a notifyError popup with the error message). This will require adding a route for setting the title on a chain event.
- [ ] The title should be limited to 140 characters, and a notifyError should be shown if any longer title is provided.
- [ ] Add a title column on the Chain Entity db table.

If the treasury proposal’s title has not been set, it should be in a lighter color, and have the title
* “No title set”

Otherwise it should have whatever title has been set.

The individual columns for treasury proposals should change to accommodate the title:
* “Value” should become “Proposed Spend”
* “Recipient” should be added after “Value”
* “Bond” should remain as is
* “Author” should remain as is
* Width of the four columns should be proportionally 1:2:1:2

### PR DESCRIPTION
**Pull request by gdjohnson**
[Link to Notion spec.](https://www.notion.so/hicommonwealth/Chain-Entity-Authors-Titles-13-2a0389729cb544c79ab44258b4fd30f8)

Closes #840.

## Description

This branch extends proposal title-setting functionality from admins/mods to proposal authors. The db and models have already been configured in previous pulls to allow storage of author and title data; this ensures that entity archival includes author information, and that the title-setting UI/title-setting endpoint allow authors to update titles.

Note: scope has changed somewhat since ticket inception; referenda do not have titles, but council motions do.

## How has this been tested?

I've tested setting/fetching on both active and completed treasury and democracy proposals, with several different author/addr combos.

Wasn't quite sure how/if to proceed on unit tests, since it falls in a weird in-between of events/api + have a pretty full plate of projects... If this is something that is high-priority I'll start exploring a best approach.

## Have proper tags been added (for bug, enhancement, breaking change)?
- [x] yes

## Does this PR affect any server routes?
- [ ] yes, and they are tested: [enter the % coverage here]
- [x] yes, and they are not tested: [enter the % coverage here]
- [ ] yes, but I did not run tests
- [ ] no

### CODE REVIEW

File: client/scripts/views/pages/view_proposal/index.ts [hunk 1]
```diff
@@ -93,10 +93,12 @@ const ProposalHeader: m.Component<{
 
     // Original posters have full editorial control, while added collaborators
     // merely have access to the body and title
-    const isAuthor = (app.user.activeAccount?.address === proposal.author
-          && app.user.activeAccount?.chain.id === (proposal as OffchainThread).authorChain);
+    const { activeAccount } = app.user;
+    const authorChain = (proposal instanceof OffchainThread) ? proposal.authorChain : app.activeId();
```
Comment (raykyri) [inline client/scripts/views/pages/view_proposal/index.ts:97]: we should just add `chain` to Proposal object in the future, but this seems fine for now

File: server/routes/updateChainEntityTitle.ts [hunk 2]
```diff
@@ -15,22 +15,24 @@ const updateChainEntityTitle = async (models, req: Request, res: Response, next:
...
-      address_id: { [Op.in]: userOwnedAddressIds, },
-      permission: { [Op.in]: ['admin', 'moderator'] },
-    }
-  });
-  const role = roles.find((r) => {
-    return r.offchain_community_id === entity.community || r.chain_id === entity.chain;
-  });
-  if (!role) return next(new Error(Errors.NotAdminOrOwner));
-  // }
+  if (!userOwnedAddresses.includes(entity.author)) {
```
Comment (raykyri) [inline server/routes/updateChainEntityTitle.ts:24]: This should check that (address matches && chain matches).

File: server/routes/updateChainEntityTitle.ts [hunk 3]
```diff
@@ -15,22 +15,24 @@ const updateChainEntityTitle = async (models, req: Request, res: Response, next:
...
+    const roles = await models.Role.findAll({
+      where: {
+        address_id: { [Op.in]: userOwnedAddressIds, },
+        permission: { [Op.in]: ['admin', 'moderator'] },
+      }
+    });
+    const role = roles.find((r) => {
+      return r.offchain_community_id === entity.community || r.chain_id === entity.chain;
+    });
```
Comment (raykyri) [inline server/routes/updateChainEntityTitle.ts:33]: Entities can't have communities.

**Review by raykyri — CHANGES_REQUESTED** [review]
Minor changes requested.

**Comment by gdjohnson** [comment]
Requested fixes pushed 👍

**Pull request merged**